### ISSUE TITLE
combine: support seeds and dyes

### ISSUE THREAD

**Issue by myk002** [DFHack/dfhack#2957]
seeds and dye that are in bags (possibly also in barrels) can and should also be combined

**Comment by myk002** [comment]
from u/daggerdragon

Level 43 has two main seed stockpiles - the feeder is SEEDS in the center top above all the flooded farmland and SEEDS: useless in the lower left corner.

When the game first loads, the useless seeds pile has 3-4 bags of 2-3 apple seeds that really should be in one bag.

Additionally, if you select everything in that 5x5 stockpile and autodump it in the middle (and unforbid), then run the game for a few seconds, the dwarves will stuff everything into a few bags or shuffle them off somewhere else or Armok knows what...

**Comment by myk002** [comment]
@silverflyone would you like to take this?

**Comment by silverflyone** [comment]
Sure

**Comment by Codedependent77** [comment]
Is combining individual bolts back into piles of 25 (or whatever) something Combine could do?

**Comment by silverflyone** [comment]
I'll look into it.

**Comment by Codedependent77** [comment]
If it does turn out to be possible, being able to limit which qualities (and possibly materials) of bolts combine would be a great addition - ideally we'd like stacks of 25 masterwork bolts for use and everything else allowed to stay individual so we can melt those down (250% efficiency for a single bolt vs 30% for a stack).

**Comment by flarejoltporeon** [comment]
It's definitely possible to add ammo to the list of what combine can work with (I modified my copy of the script to get that functionality, but there's someone now looking into this I'll pass the code on rather than do my own pull request (it's intimidating!))

@silverflyone 
I added the following to line 20, just below the ['all'] map entry: 
   ` ['ammo'] = {[df.item_type.AMMO]={type_id=df.item_type.AMMO, type_name='AMMO',type_caste=false,max_stack_size=DEF_MAX}},`

I didn't filter it at all, it takes any bolt in the stockpile when I use `combine here` (combine all freezes in my current fort, I think due to meal sizes, so I only ever go one stockpile at a time). Not tested with `combine all`.

**Comment by myk002** [comment]
@flarejoltporeon Freezing your fort with `combine all` is a serious bug! could you upload your fort and mods somewhere to the cloud and post a link to them here so we can investigate?

**Comment by silverflyone** [comment]
@flarejoltporeon Yes it would help for debugging if you could share your save, and all your mods you use.

**Comment by flarejoltporeon** [comment]
@myk002 and @silverflyone 

You can find a save at https://dffd.bay12games.com/file.php?id=16545 (Please ignore it says it's for the 9,000 pants issue - `combine all` still crashes it, and I reverted to the unmodified version of combine just to make sure. 

I was always just assuming I'd done something wrong in the past with my commands and borked it entirely.

Mod List (Zip attached to this comment): 
See-Through Smoothing Designations
clinodev's dry mines (no aquifers)
Dwarven Weapons Fixed 2.0
Interface Tweaks 1.3
Let Dwarven Women Have Beards 1.4
No Attribute/Skill Rust 1.0.0
Rounded Hills 1.6

[Mods.zip](https://github.com/DFHack/dfhack/files/11136664/Mods.zip)

**Comment by silverflyone** [comment]
@myk002 Discussion for seeds and dyes. Thoughts?

**Seeds**
Understanding: Seeds are non stackable (I've confirmed this through testing: if they have a stack size > 1 then the dwarves plant the whole stack = not expected behaviour). Seeds are generally contained in a bag, but can be dumped on the ground; meaning they could exist by themselves on the ground in a stockpile. I haven't seen any seeds in a barrel without a bag, but assume this may be possible. Most people store seeds in bags in stockpiles rather than in barrels; assume this is due to problems with planting functions. There is no DF program defined limit for how many seeds can exist in a bag/container; I've seen hundreds in a single bag. Seeds don't have a weight.

Combine approach: Filter by seed. Will not stack seeds since they are marked not stackable; i.e. stack_size is undefined. Will combine seeds contained in a bag. Will loop through the containers (bags) and detach/move the seeds so that comparable seeds are contained in bag. A MAX_CON of 500 will be used to limit seed count per bag.  Bags in stockpiles will be prioritized ahead of bags in barrels (optional - need to figure this out).

**Dyes**
Understanding:  dyes are a powder_misc type. Powder_misc includes material: dye (420), flour/sugar (421), plaster, quicklime and sand. Dye, flour/sugar and quicklime can be stacked, but plaster and sand cannot be stacked; even though is_stackable is true for powder_misc type.

Combine approach: Ignore plaster and sand, since these are non trivial cases and don't stack like the others. Will filter by dye, flour, sugar and quicklime = use type, mat type and mat index as the matching criteria for these items - need to figure out an efficient way to search for these, thinking to use a flag for powder types to specify additional search criteria.

**Comment by silverflyone** [comment]
@flarejoltporeon @myk002 I've looked into the freeze. It is an unhandled exception in Buildings.cpp. It has its own issue: https://github.com/DFHack/dfhack/issues/3159.

**Comment by myk002** [comment]
> Bags in stockpiles will be prioritized ahead of bags in barrels (optional - need to figure this out)

Since the TSK issue has been (reportedly -- I haven't personally tested this) fixed (pre-steam, only one dwarf could use a container at once, meaning seeds in a container couldn't be planted in parallel), I'd say we can simplify a little bit. All seeds of one type can go into a single bag without limits (there is already a game-wide per-seed limit that will act as a throttle), and there's no need to prefer bags in stockpiles over bags in barrels.

You can use `math.huge` as the default seed stack size.

**Comment by silverflyone** [comment]
> > Bags in stockpiles will be prioritized ahead of bags in barrels (optional - need to figure this out)
> 
> 
> 
> Since the TSK issue has been (reportedly -- I haven't personally tested this) fixed (pre-steam, only one dwarf could use a container at once, meaning seeds in a container couldn't be planted in parallel), I'd say we can simplify a little bit. All seeds of one type can go into a single bag without limits (there is already a game-wide per-seed limit that will act as a throttle), and there's no need to prefer bags in stockpiles over bags in barrels.
> 
> 
> 
> You can use `math.huge` as the default seed stack size.

PR raised for seeds and dyes: https://github.com/DFHack/scripts/pull/676. Appreciate any feedback on this one.

**Comment by myk002** [comment]
Fixed by DFHack/scripts#681
